test/cluster/test_client_routes.py

```diff
@@ -159,3 +160,115 @@ async def fail_req(f):
+    """
+
+    servers = await manager.servers_add(2, cmdline=['--smp=2'])
+    cql, hosts = await manager.get_ready_cql(servers)
+
+    received_events = []
+    setup_events_test(cql, received_events)
+
+    await manager.api.client.post("/v2/client-routes", host=servers[0].ip_addr, json=[generate_client_routes_entry(0)])
+
+    await wait_for_expected_event_num(1, received_events)
+    assert received_events[0]["change_type"] == "UPDATE_NODES"
+    assert received_events[0]["connection_ids"] == [generate_connection_id(0)]
+    assert received_events[0]["host_ids"] == [generate_host_id(0)]
+
+    await manager.api.client.post("/v2/client-routes", host=servers[0].ip_addr, json=[
+        generate_client_routes_entry(1),
+        generate_client_routes_entry(2),
+    ])
+    await wait_for_expected_event_num(2, received_events)
+    assert received_events[1]["change_type"] == "UPDATE_NODES"
+    assert received_events[1]["connection_ids"] == [generate_connection_id(1), generate_connection_id(2)]
+    assert received_events[1]["host_ids"] == [generate_host_id(1), generate_host_id(2)]
+
+    await manager.api.client.delete("/v2/client-routes", host=servers[0].ip_addr, json=[generate_client_routes_entry(0)])
+    await wait_for_expected_event_num(3, received_events)
+    assert received_events[2]["change_type"] == "UPDATE_NODES"
+    assert received_events[2]["connection_ids"] == [generate_connection_id(0)]
+    assert received_events[2]["host_ids"] == [generate_host_id(0)]
+
+@pytest.mark.asyncio
+@skip_mode("release", "error injections are not supported in release mode")
+async def test_client_routes_snapshot_transfer(request, manager: ManagerClient):
+    """
+    This test verifies that client routes change events are sent when client_routes
+    data is propagated via snapshot transfer:
+      1. Create a 3-node cluster.
+      2. Enable `wait_transfer_snapshot` error injection on one node, and stop it.
+      3. Change client routes with a POST request on other nodes, and trigger a snapshot.
+      4. Start the stopped node, and send a message to stop waiting on `wait_transfer_snapshot`.
+      5. Verify that an event was sent.
+    """
+    servers = await manager.servers_add(3, cmdline=['--smp=2'])
+    cql, hosts = await manager.get_ready_cql(servers)
+    server_to_restart = servers[2]
+    error_to_inject = "wait_transfer_snapshot"
+
+    await manager.server_update_config(server_to_restart.server_id, "error_injections_at_startup", [error_to_inject])
+    await manager.server_stop(server_to_restart.server_id)
+
+    await manager.api.client.post("/v2/client-routes", host=servers[0].ip_addr, json=[generate_client_routes_entry(1)])
+    await wait_for_expected_client_routes_size(cql, 1)
+    await trigger_snapshot(manager, servers[0])
+
+    await manager.server_start(server_to_restart.server_id)
+    cql = await manager.get_cql_exclusive(server_to_restart)
+    await wait_for_expected_client_routes_size(cql, 0)
```

Comment:
This check looks unreliable.  It could happen that we read from `client_routes` too quickly, before a snapshot/command is (unexpectedly) applied. I would wait for a log printed after hitting wait_transfer_snapshot.

We should also verify that this test does what we expect. The test could be broken (by e.g. misusing trigger_snapshot) and `client_routes` could be unexpectedly updated in a group 0 command, not snapshot. We should check that this update really happens in transfer_snapshot. An easy way to do this is by adding a trace log there for non-empty client_routes updates and looking for it in the test.